cobra/util/solver.py

```diff
@@ -0,0 +1,170 @@
+           for match in dir(optlang) if "_interface" in match}
+"""
+Defines all the solvers that were found in optlang.
+"""
+
+
+def interface_to_str(interface):
+    """Give a string representation for an optlang interface.
+
+    Parameters
+    ----------
+    interface: string
+    Full name of the interface in optlang or cobra representation.
+    For instance 'optlang.glpk_interface' or 'optlang-glpk'.
+
+    Returns
+    -------
+    name: string
+    The name of the interface as a string
+    """
+    return re.sub(r"optlang.|.interface", "", interface)
+
+
+def get_solver_name(mip=False, qp=False):
+    """Selects a solver for a given optimization problem in a reproducible
+    manner.
+
+    Parameters
+    ----------
+    mip: string
+    Does the solver require mixed integer linear programming capabilities?
+    qp: string
+    Does the solver require quadratic programming capabilities?
+
+    Returns
+    -------
+    name: string
+    The name of feasible solver.
+
+    Notes
+    -----
+    Raises SolverNotFound if a suitable solver is not found.
+    """
+    if len(solvers) == 0:
+        raise SolverNotFound("no solvers installed")
+    # glpk only does lp, not qp. Gurobi and cplex are better at mip
+    mip_order = ["gurobi", "cplex", "mosek", "glpk"]
+    lp_order = ["glpk", "cplex",  "gurobi", "mosek", "scipy"]
+    qp_order = ["gurobi", "cplex", "mosek"]
+
+    if mip is False and qp is False:
+        for solver_name in lp_order:
+            if solver_name in solvers:
+                return solver_name
+        # none of them are in the list order - so return the first one
+        return list(solvers)[0]
+    elif qp:  # mip does not yet matter for this determination
+        for solver_name in qp_order:
+            if solver_name in solvers:
+                return solver_name
+        raise SolverNotFound("no qp-capable solver found")
+    else:
+        for solver_name in mip_order:
+            if solver_name in solvers:
+                return solver_name
+    raise SolverNotFound("no mip-capable solver found")
+
+
+def add_to_solver(model, what=[]):
+    """Adds variables and constraints to a Model's solver object.
+
+    Useful for variables and constraints that can not be expressed with
+    reactions and lower/upper bounds. Will integrate with the Model's context
+    manager in order to revert changes upon leaving the context.
+
+    Parameters
+    ----------
+    model: a cobra model
+    The model to which to add the variables and constraints.
+    what: list or tuple of optlang variables or constraints.
+    The variables or constraints to add to the model. Must be of class
+    `model.solver.interface.Variable` or `model.solver.interface.Constraint`.
+    """
+    context = get_context(model)
+
+    if len(what) > 0:
+        model.solver.add(what)
+        if context:
+            context(partial(model.solver.remove, what))
+
+
+def remove_from_solver(model, what=[]):
```

Comment:
right, but we could replace all calls to `model.solver.add()` with `add_to_solver`. We could even consider making it a cobra.Model method, `model.add_to_solver(self, ...)`